src/Sylius/Bundle/LocaleBundle/Form/Extension/DateExtension.php

```diff
@@ -0,0 +1,131 @@
+use Sylius\Bundle\LocaleBundle\Templating\Helper\LocaleHelper;
+use Sylius\Component\Locale\Calendars;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\Form\FormView;
+use Symfony\Component\Form\AbstractTypeExtension;
+
+/**
+ * Extension for Symfony's DateType for a localized calendar-aware date.
+ *
+ * @author Aram Alipoor <aram.alipoor@gmail.com>
+ */
+class DateExtension extends AbstractTypeExtension
+{
+    /**
+     * @var LocaleHelper
+     */
+    protected $localeHelper;
+
+    /**
+     * @param LocaleHelper $helper
+     */
+    public function __construct(LocaleHelper $helper)
+    {
+        $this->localeHelper = $helper;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        if (Calendars::GREGORIAN === $this->localeHelper->getCurrentCalendar()) {
+            // We only need to alter default behaviour when
+            // we're having a traditional calendar system.
+            return;
+        }
+
+        if ('single_text' === $options['widget']) {
+            $builder->addViewTransformer(new StringGregorianToCalendarSystemTransformer(
+                'Y-m-d',
+                $this->localeHelper
+            ));
+        } elseif ('array' === $options['input']) {
+            $builder->addViewTransformer(new ArrayGregorianToCalendarSystemTransformer(
+                $this->localeHelper
+            ));
+        }
+
+        if ('choice' === $options['widget']) {
+            // If widget is of type `choice` we need to translate choice items
+            // stored in `choices` of the field to their localized version.
+            $this->translateChoices($builder, 'year', 'Y');
+            $this->translateChoices($builder, 'month', 'm');
+            $this->translateChoices($builder, 'day', 'd');
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function finishView(FormView $view, FormInterface $form, array $options)
+    {
+        if ('single_text' === $options['widget'] &&
+            Calendars::GREGORIAN !== $this->localeHelper->getCurrentCalendar()) {
+            // Native browsers do not support different calendar systems
+            // so we have to use "text" type instead of "date" for the input.
+            // We only need to alter default behaviour when
+            // we're having a traditional calendar system.
+            $view->vars['type'] = 'text';
+        }
+    }
+
+    /**
+     * A helper method to translate choices of a date-related field
+     * to their localized version.
+     *
+     * e.g. For a year field it should
+     *      translate 2014 (gregorian) choice to to 1393 (persian)
+     *
+     * @param FormBuilderInterface $builder
+     * @param string               $fieldName
+     * @param string               $format PHP-compatible date format
```

Comment:
If you're aligning variables in docblocks, try to be consistent and do it in all your methods.